### DISCUSSION TITLE
Use email only for login/registration

### DISCUSSION THREAD

**Discussion by kipchirchirian** [codeigniter4/shield#192] [Q&A] [answered]
I need to use just the email for login/registration. What are the steps to configure or setup this?

**Comment by kenjis** [comment, marked as answer]
I've not tried, but how about this?

1. remove `username` in the config.
https://github.com/codeigniter4/shield/blob/44901bcd27d4f87535126de426a361182ad67775/src/Config/Auth.php#L216-L223
2. remove `username` in the HTML form in `src/Views/register.php`.
3. remove `username` validation rule in `RegisterController`.
4. customize the login form.

**Reply by kipchirchirian** [reply]
Although I marked this as the answer, but the steps above is just half of it and won't work. In addition to what @lonnieezell has written below, what I had to do for it to completely work is:

1. After copying the `RegisterController.php`, update the routes by passing the except option:

```php
service('auth')->routes($routes, ['except' => ['register']]);
```

2. Add routes for your customized controller
```php
$routes->post('register', '\App\Controllers\Auth\RegisterController::registerAction');
```

I've only focused on the important parts based on the question.

**Reply by MGatner** [reply]
Your way of routing will work fine but also just to note: you could just set the new controller in your **AuthRoutes.php** Config file! That's what it's there for.

**Comment by lonnieezell** [comment]
Yes, the steps that @kenjis provided will work for you. You should probably copy the views and controller to your application, though.

**Comment by genio** [comment]
This all seems like quite the amount of work to shunt onto the user. Requiring the user to copy-paste the `RegisterController` and force the `username` to be `null` seems overkill.
Could that could be more easily done by removing it from the `getRegistrationRules()` if the Auth config contains `$registrationType = 'emailOnly';` or some such other simple thing?

**Reply by datamweb** [reply]
> This all seems like quite the amount of work to shunt onto the user. Requiring the user to copy-paste the `RegisterController` and force the `username` to be `null` seems overkill.

There is no need for this. Proceed as follows.


Copy **vendor\codeigniter4\shield\src\Views\register.php** to **app\Views\register.php**

Delete this code:
```php
                    <!-- Username -->
                    <div class="form-floating mb-4">
                        <input type="text" class="form-control" id="floatingUsernameInput" name="username" inputmode="text" autocomplete="username" placeholder="<?= lang('Auth.username') ?>" value="<?= old('username') ?>" required>
                        <label for="floatingUsernameInput"><?= lang('Auth.username') ?></label>
                    </div>
```

Edit app\Config\Auth.php

```php
    public array $views = [
        'register'                    => '\App\Views\register',

    ];
```

Add app/Config/Validation.php
```php
    public $registration = [

        'email' => [
            'label' => 'Auth.email',
            'rules' => [
                'required',
                'max_length[254]',
                'valid_email',
                'is_unique[auth_identities.secret]',
            ],
        ],
        'password' => [
            'label' => 'Auth.password',
            'rules' => 'required|max_byte[72]|strong_password[]',
            'errors' => [
                'max_byte' => 'Auth.errorPasswordTooLongBytes'
            ]
        ],
        'password_confirm' => [
            'label' => 'Auth.passwordConfirm',
            'rules' => 'required|matches[password]',
        ],
    ];
```



> Could that could be more easily done by removing it from the `getRegistrationRules()` if the Auth config contains `$registrationType = 'emailOnly';` or some such other simple thing?

I think ordering is simple enough and we don't need to add a new option.

**Reply by genio** [reply]
Thank you! That reply was very helpful. I was unaware of the ability to completely override the validation rules in the Validation config file.
